src/Control/Distributed/Process/Node/RegistryAgent.hs

```diff
@@ -0,0 +1,81 @@
+---- Portability :  non-portable (requires concurrency)
+----
+---- This module provides registry monitoring agent, implemented as
+---- a /distributed-process Management Agent/. Once 'node' starts it run this
+---- agent. Agent will monitor every remote process that was added to the
+---- local registry. So node will receive Died notification and remove process
+---- from registry in case if target process is dead.
+----
+-------------------------------------------------------------------------------
+
+module Control.Distributed.Process.Node.RegistryAgent
+    ( registryMonitorAgent
+    ) where
+
+import Control.Distributed.Process.Management
+import Control.Distributed.Process.Internal.Types
+import Control.Distributed.Process.Internal.Primitives
+import Data.Foldable (forM_)
+#if MIN_VERSION_containers(5,0,0)
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+#else
+import Data.Map (Map)
+import qualified Data.Map as Map
+#endif
+
+registryMonitorAgentId :: MxAgentId
+registryMonitorAgentId = MxAgentId "service.registry.monitoring"
+
+-- | Registry monitor agent. Agent listens for 'MxRegistered' and 'MxUnRegistered'
+-- events and tracks all remote 'ProcessId's that are stored in registry.
+-- In case if a new remote process is registered node starts monitoring that
+-- process until it dies or unregistered from node. Agent keeps ref counting
+-- for all process as a result if process is stored in registry multiple times
+-- it will not be unmonitored unless all references will be removed.
+registryMonitorAgent :: Process ProcessId
+registryMonitorAgent = do
+    nid <- getSelfNode
+    mxAgent registryMonitorAgentId initState
+        [ mxSink $ \(ProcessMonitorNotification _ pid _) -> do
+            mxUpdateLocal (Map.delete pid)
+            mxReady
+        , mxSink $ \ev ->
+            let act = case ev of
+                    MxRegistered pid _
+                        | processNodeId pid /= nid -> do
+                        hm <- mxGetLocal
+                        case pid `Map.lookup` hm of
+                            Nothing -> do
+                                mon <- liftMX $ monitor pid
+                                mxUpdateLocal (Map.insert pid (mon, 1))
+                            Just (mref,i)  ->
+                                let !i' = succ i
+                                in mxUpdateLocal (Map.insert pid (mref, i'))
+                    MxUnRegistered pid _
+                        | processNodeId pid /= nid -> do
+                        hm <- mxGetLocal
+                        forM_ (pid `Map.lookup` hm) $ \(mref, i) ->
+                           let !i' = pred i
+                           in if i' == 0
+                              then do liftMX $ unmonitorAsync mref
+                                      mxSetLocal $! pid `Map.delete` hm
+                              else mxSetLocal $ Map.insert pid (mref,i') hm
+                    _ -> return ()
```

Comment:
Unfortunately both of those will not work here.

1. `insertLookupWithKey` will not work as I have to have `a` that is `(MonitorRef, Int)` and I need to create `MonitorRef` only if there is no value. 

2. `updateLookupWithKey` returns either old value if value was deleted or changed value, so I have no chance
   to understand what happens as value `(mon,1)` can refer either to changed value or to removed.